### ISSUE TITLE
Discovery of approach for background sync for dataset uploads

### ISSUE THREAD

**Issue by ckrountree** [OpenNeuroOrg/openneuro#1344]

**Comment by nellh** [comment]
Here's what I've found after prototyping a bit, we cannot use background sync due to the size of files we're working with. The thresholds adopted by browsers are a small portion of free disk space, restrictions vary depending on browser but it's somewhere in the range of a few gigabytes to tens of gigabytes. Since the current uploader can do really any size and the web downloader works in the range of 5-10GB, this is a regression and won't be useful given the current implementations.

Background fetch is almost identical to fetch but offers most of the advantages of background sync. An implementation is stable in Chrome 74. Not much has to change to support this as an optional transport for upload or download requests, falling back to fetch for browsers which do not support the newer API (Firefox for now). Background fetch should let us more reliably resume uploads without user interaction, making most failures more transparent to the user. Just implementing background fetch would improve upload reliability but would have a limited improvement on downloads.

To solve download transfer issues, we should move away from compressing the dataset in the service worker. This approach is fine for small datasets on fast computers, but it suffers from 32-bit integer bugs and performance degradation in many scenarios. Since we are limited to streaming the resulting zip archive, there is no opportunity to resume if a file download fails mid-transfer. Another upcoming API just entered origin trial in Chrome 78, the native file API. The native file API solves this issue by requesting permission to write directly to the user's filesystem. Combined with background fetch, this gives us a durable transport and the ability to retry individual files or sets of files as needed without forcing the user to initiate each file download independently. This is less standardized and limits how easily we can maintain a fallback implementation.

Short version:

1. We should use background fetch to move in both directions.
1. We should use the native filesystem API in both directions.
1. There is some time until both APIs are broadly available and there is some risk with the native file API changing significantly, so they cannot replace the current solutions immediately.
